qiskit/ignis/experiments/calibration/instruction_data/gate_and_schedules_definition.py

```diff
@@ -0,0 +1,140 @@
+from qiskit.circuit import ParameterExpression, Gate, Parameter
+from qiskit.pulse import Play, Schedule, ControlChannel, ParametricPulse
+from qiskit.pulse.channels import PulseChannel
+
+
+class InstructionsDefinition:
+    """
+    Class to track calibrated instructions for the calibration module of Ignis.
+
+    The class allows users to define single-pulse schedules as quantum circuits and
+    then leverage the quantum circuit compose operation to compose more complex
+    circuits from basic circuits.
+    """
+
+    def __init__(self):
+
+        # Dict to store the instructions we are calibrating
+        self._instructions = {}
+
+    @property
+    def instructions(self) -> Dict[Tuple[str, Iterable], QuantumCircuit]:
+        """Return the instructions as a dict."""
+        return self._instructions
+
+    def get_instruction_template(self, name: str, qubits: Iterable) -> QuantumCircuit:
+        """
+        Args:
+            name: Name of the instruction to get.
+            qubits: Set of qubits to which the instruction applies.
+
+        Returns:
+            A QuantumCircuit with parameters in it.
+        """
+        return self._instructions[(name, qubits)]
+
+    def get_instruction(self, name: str, qubits: Iterable,
+                        free_parameters: List = None) -> QuantumCircuit:
+        """
+        Returns the QuantumCircuit of the instruction where all parameters aside for those
+        explicitly specified are bound to their calibration.
+
+        Args:
+            name: name of the instruction to retrive
+            qubits: qubits to which the instruction applies.
+            free_parameters: Parameters that should be left unbound. If None is specified then
+                all parameters will be bound to their calibrated values.
+        """
+        # TODO Get the parameters
+
+        # TODO Get their values
+
+        # TODO return circ with parameters assigned
+        raise NotImplemented
+
+    def create_basic_instruction(self, name: str, duration: int, pulse_envelope: ParametricPulse,
+                                 channel: PulseChannel, params = None):
+        """
+        Creates a basic instruction in the dictionary.
+
+        Args:
+            name:
+            duration: The duration of the pulse schedule.
+            pulse_envelope: The callable pulse envelope. Its parameters are given by
+                **params after duration is added to it.
+            channel: The channel to which we apply the basic instruction.
+        """
+        # Avoids having the user manage the parameters.
+        if not params:
+            params = self._basic_inst_parameters(name, channel, pulse_envelope)
+
+        gate = Gate(name=name, num_qubits=1, params=[_ for _ in params.values()])
+
+        if duration:
+            params['duration'] = duration
+
+        schedule = Schedule(name=name)
+        schedule = schedule.insert(0, Play(pulse_envelope(**params), channel))
```

Comment:
How can we define Z gate? there is no parametric pulse for Z because they are frame changes, but we need to know how Z is defined to generate basis gates. We cannot do sideband with this approach (because Qiskit pulse instruction doesn't have sideband so we need schedule to combine the instruction with sideband).

Reply:
Yes, this is still lacking here. We could perhaps explicitly create a `Z` gate by adding an function for this in addition to `create_basic_instruction` e.g. `create_z_instruction` or use `ShiftPhase`.

Reply:
Another question. To construct a basis gates, we need frame sync among some channels. Can we implement this with current approach?

Reply:
E.g. if I do a frame change on `d0` we also need to do frame changes on the corresponding `u%i` channels? I think that should be possible by defining the `Z` gate as a schedule that does `fc` on the qubit and frame changes on the corresponding `u` channels.

Reply:
how can we define frame changes? is this a part of basic instruction? or frame change can be composite?

Reply:
I added a definition of a `Z` gate using ShiftPhase.